File: CPP/No787.cc
/**
 * Created by Xiaozhong on 2020/12/14.
 * Copyright (c) 2020/12/14 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <string>
#include <queue>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>> &flights, int src, int dst, int K) {
        vector<vector<int>> graph(n, vector<int>(n));
        for (const vector<int> &flight : flights) {
            graph[flight[0]][flight[1]] = flight[2];
        }
        unordered_map<int, int> best;
        struct cmp {
            bool operator()(const vector<int> &lhs, const vector<int> &rhs) {
                return lhs.front() < rhs.front();
            }
        };
        priority_queue<vector<int>, vector<vector<int>>, cmp> pq;
        pq.push({0, 0, src});
        while (!pq.empty()) {
            vector<int> info = pq.top();
            pq.pop();
            int cost = info[0], k = info[1], place = info[2];
            if (k > K + 1 || cost > best.count(k * 1000 + place) ? best[k * 1000 + place] : INT32_MAX) continue;
            if (place == dst) return cost;
            for (int nei = 0; nei < n; ++nei)
                if (graph[place][nei] > 0) {
                    int newcost = cost + graph[place][nei];
                    if (newcost < best.count((k + 1) * 1000 + nei) ? best[(k + 1) * 1000 + nei] : INT32_MAX) {
                        pq.push({newcost, k + 1, nei});
                        best[(k + 1) * 1000 + nei] = newcost;
                    }
                }
        }
        return -1;
    }
};

File: Java/src/No787.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.PriorityQueue;

public class No787 {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        int[][] graphs = new int[n][n];
        for (int i = 0; i < flights.length; ++i) {
            graphs[flights[i][0]][flights[i][1]] = flights[i][2];
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        pq.offer(new int[]{src, 0, K + 1});
        while (!pq.isEmpty()) {
            int[] front = pq.poll();
            int from = front[0];
            int price = front[1];
            int k = front[2];
            if (from == dst) return price;
            if (k > 0) {
                for (int i = 0; i < n; ++i) {
                    if (graphs[from][i] > 0)
                        pq.offer(new int[]{i, price + graphs[from][i], k - 1});
                }
            }
        }
        return -1;
    }
}
